src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsassessforearlyreleaseapi/entity/AssessmentStatus.kt

```diff
@@ -282,19 +135,466 @@ enum class AssessmentStatus {
+      }
+      on<AssessmentLifecycleEvent.StartAddressChecks> {
+        transitionTo(AssessmentState.AddressAndRiskChecksInProgress)
+      }
+      on<AssessmentLifecycleEvent.SubmitForDecision> {
+        transitionTo(AssessmentState.AwaitingPreDecisionChecks)
+      }
+      on<AssessmentLifecycleEvent.Approve> {
+        transitionTo(AssessmentState.Approved)
+      }
+      on<AssessmentLifecycleEvent.CompleteAddressChecks> {
+        transitionTo(AssessmentState.AwaitingPreReleaseChecks)
+      }
+      on<AssessmentLifecycleEvent.ReleaseOnHDC> {
+        transitionTo(AssessmentState.PassedPreReleaseChecks)
+      }
+      on<AssessmentLifecycleEvent.FailAddressChecks> {
+        transitionTo(AssessmentState.AddressUnsuitable)
+      }
+      on<AssessmentLifecycleEvent.Refuse> {
+        transitionTo(AssessmentState.Refused)
+      }
+      on<AssessmentLifecycleEvent.OptOut> {
+        transitionTo(AssessmentState.OptedOut(this.status))
+      }
+    }
+
+    state<AssessmentState.Postponed> {
+      on<AssessmentLifecycleEvent.SubmitForDecision> {
+        transitionTo(AssessmentState.AwaitingPreDecisionChecks)
+      }
+      on<AssessmentLifecycleEvent.Approve> {
+        transitionTo(AssessmentState.Approved)
+      }
+      on<AssessmentLifecycleEvent.CompleteAddressChecks> {
+        transitionTo(AssessmentState.AwaitingPreReleaseChecks)
+      }
+      on<AssessmentLifecycleEvent.ReleaseOnHDC> {
+        transitionTo(AssessmentState.PassedPreReleaseChecks)
+      }
+      on<AssessmentLifecycleEvent.Timeout> {
+        transitionTo(AssessmentState.TimedOut)
+      }
+    }
+
+    state<AssessmentState.OptedOut> {
+      on<AssessmentLifecycleEvent.PassEligibilityAndSuitability> {
+        transitionTo(AssessmentState.EligibleAndSuitable)
+      }
+      on<AssessmentLifecycleEvent.SubmitForAddressChecks> {
+        transitionTo(AssessmentState.AwaitingAddressAndRiskChecks)
+      }
+      on<AssessmentLifecycleEvent.StartAddressChecks> {
+        transitionTo(AssessmentState.AddressAndRiskChecksInProgress)
+      }
+      on<AssessmentLifecycleEvent.SubmitForDecision> {
+        transitionTo(AssessmentState.AwaitingPreDecisionChecks)
+      }
+      on<AssessmentLifecycleEvent.Approve> {
+        transitionTo(AssessmentState.Approved)
+      }
+      on<AssessmentLifecycleEvent.CompleteAddressChecks> {
+        transitionTo(AssessmentState.AwaitingPreReleaseChecks)
+      }
+      on<AssessmentLifecycleEvent.ReleaseOnHDC> {
+        transitionTo(AssessmentState.PassedPreReleaseChecks)
+      }
+      on<AssessmentLifecycleEvent.FailAddressChecks> {
+        transitionTo(AssessmentState.AddressUnsuitable)
+      }
+      on<AssessmentLifecycleEvent.Refuse> {
+        transitionTo(AssessmentState.Refused)
+      }
+      on<AssessmentLifecycleEvent.Timeout> {
+        transitionTo(AssessmentState.TimedOut)
+      }
+      on<AssessmentLifecycleEvent.OptBackIn> {
+        transitionTo(this.previousStatus.toState())
```

Comment:
As mentioned above, note that here we can't currently transition back to a state that also have a previous state.

We'd need to access the state history